### PR TITLE
[Github] Changelog > Releases Action

### PR DESCRIPTION
**Pull request by MickLesk**
## Description
Creates a release based on the changelog.

There should actually only be the CHANGELOG.md in the RELEASE and no source code.

It is created daily at 23.59 UTC if there have been changes to the changelog.md.

## Type of change
Please check the relevant option(s):

- [ ] Bug fix (non-breaking change that resolves an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (a fix or feature that would cause existing functionality to change unexpectedly)
- [ ] New script (a fully functional and thoroughly tested script or set of scripts.)

## Prerequisites
The following efforts must be made for the PR to be considered. Please check when completed:
- [x] Self-review performed (I have reviewed my code, ensuring it follows established patterns and conventions)
- [x] Testing performed (I have tested my changes, ensuring everything works as expected)
- [x] Documentation updated (I have updated any relevant documentation)

@havardthom:
Please check this carefully. I have built an action to the best of my knowledge and belief using the Github Docu. You are also welcome to make direct changes yourself.

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -0,0 +1,67 @@
+name: Create new release
+
+on:
+  schedule:
+    # Runs "At 23:59 every night" (UTC)
+    - cron: '59 23 * * *'
+
+jobs:
+  fetch-changelog:
```
Comment (havardthom) [inline .github/workflows/main.yml:9]: Can remove this job. `CHANGELOG.md` is available after checkout in `create-new-release` job

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    needs: fetch-changelog
+    permissions:
+      contents: write
+      pull-requests: write
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+      - name: Download CHANGELOG.md
```
Comment (havardthom) [inline .github/workflows/main.yml:32]: This step is not needed

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    runs-on: ubuntu-latest
+    needs: fetch-changelog
+    permissions:
+      contents: write
+      pull-requests: write
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
```
Comment (havardthom) [inline .github/workflows/main.yml:31]:
```suggestion
```
Don't need full commit history

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        with:
+          name: changelog
+          path: CHANGELOG.md
+
+  create-new-release:
+    runs-on: ubuntu-latest
+    needs: fetch-changelog
+    permissions:
+      contents: write
+      pull-requests: write
```
Comment (havardthom) [inline .github/workflows/main.yml:26]:
```suggestion
```
pull request permissions not needed

File: .github/workflows/main.yml [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+      - name: Parse CHANGELOG for today's entries and create a new tag
+        if: steps.verify-diff.outputs.changed == 'true'
+        run: |
+          # Extract the date of today from the CHANGELOG and filter relevant entries
+          TODAY=$(date -u +'%Y-%m-%d')
+          # Find the lines in the CHANGELOG corresponding to today’s date
+          LINES=$(nl -v0 -ba CHANGELOG.md | grep "## $TODAY" - | head -n 2 | awk '{print $1}')
+          DESCRIPTION=$(cat CHANGELOG.md | head -n $(echo $LINES | cut -d' ' -f2) | tac | head -n -$(echo $LINES | cut -d' ' -f1) | head -n -1 | tac)
+
+          # Git commit configuration
```
Comment (havardthom) [inline .github/workflows/main.yml:49]: this section should be removed

We only need to see if CHANGELOG section corresponding to today’s date exist, and if it does, use it as release body. No need to any pushing or git diffing

File: .github/workflows/main.yml [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+          git push -u origin main
+          git tag -a "$TODAY" -m "$DESCRIPTION"
+          git push --tags
+
+          echo GITHUB_TAG_RELEASE_VERSION=$TODAY >> $GITHUB_ENV
+          echo "$DESCRIPTION" > RELEASE_CHANGELOG.md
+
+      - name: Create release
+        if: steps.verify-diff.outputs.changed == 'true'
+        uses: ncipollo/release-action@v1
```
Comment (havardthom) [inline .github/workflows/main.yml:63]: We don't want to use third party actions. Use github cli instead

Comment (MickLesk) [reply .github/workflows/main.yml:63]: Do you happen to have time to customize it? It's hanging in here in the branch. So you can adjust it yourself at any time. 

I probably won't be back on the PC until tomorrow. :-)

Comment (havardthom) [reply .github/workflows/main.yml:63]: I can have a look later today

**Comment by havardthom** [comment]
Opened a new PR here https://github.com/community-scripts/ProxmoxVE/pull/462, since I had to test it on my fork. Closing this

**Pull request closed without merging**